### ISSUE TITLE
high VRAM usage with large batches or resolutions

### ISSUE THREAD

**Issue by sdexpanim** [cubiq/ComfyUI_IPAdapter_plus#22]
Hi, thanks for this great repo, 

I have been running into vram issue on my 4090 when trying to use IPadapter with anything greater than batch size 16 at 512x768. 
This is mostly an issue in Animate diff, where it would be nice to generate larger batches or even at larger sizes. 

Wondering if this is expected memory usage for IPadapter? If so, please feel free to close this issue.

Here's my simple workflow for reproducing the issue and the error message for out of memory below:

![Screenshot 2023-09-30 223712](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/144590036/4a4bb037-0224-46e5-be24-065d6c5d31f4)

![ComfyUI_temp_jxprh_00008_](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/144590036/e60b3caa-ba8f-43c9-9c6f-9be67aee7108)

```
Error occurred when executing KSampler:

Allocation on device 0 would exceed allowed memory. (out of memory)
Currently allocated : 21.72 GiB
Requested : 720.00 MiB
Device limit : 23.65 GiB
Free (according to CUDA): 29.69 MiB
PyTorch limit (set by user-supplied memory fraction)
: 17179869184.00 GiB

File "/workspace/ComfyUI/execution.py", line 152, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
File "/workspace/ComfyUI/execution.py", line 82, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
File "/workspace/ComfyUI/execution.py", line 75, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
File "/workspace/ComfyUI/nodes.py", line 1236, in sample
return common_ksampler(model, seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise)
File "/workspace/ComfyUI/nodes.py", line 1206, in common_ksampler
samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
File "/workspace/ComfyUI/custom_nodes/ComfyUI-AnimateDiff-Evolved/animatediff/sampling.py", line 107, in animatediff_sample
return orig_comfy_sample(model, *args, **kwargs)
File "/workspace/ComfyUI/comfy/sample.py", line 97, in sample
samples = sampler.sample(noise, positive_copy, negative_copy, cfg=cfg, latent_image=latent_image, start_step=start_step, last_step=last_step, force_full_denoise=force_full_denoise, denoise_mask=noise_mask, sigmas=sigmas, callback=callback, disable_pbar=disable_pbar, seed=seed)
File "/workspace/ComfyUI/comfy/samplers.py", line 785, in sample
return sample(self.model, noise, positive, negative, cfg, self.device, sampler(), sigmas, self.model_options, latent_image=latent_image, denoise_mask=denoise_mask, callback=callback, disable_pbar=disable_pbar, seed=seed)
File "/workspace/ComfyUI/comfy/samplers.py", line 690, in sample
samples = sampler.sample(model_wrap, sigmas, extra_args, callback, noise, latent_image, denoise_mask, disable_pbar)
File "/workspace/ComfyUI/comfy/samplers.py", line 630, in sample
samples = getattr(k_diffusion_sampling, "sample_{}".format(sampler_name))(model_k, noise, sigmas, extra_args=extra_args, callback=k_callback, disable=disable_pbar, **extra_options)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
return func(*args, **kwargs)
File "/workspace/ComfyUI/comfy/k_diffusion/sampling.py", line 137, in sample_euler
denoised = model(x, sigma_hat * s_in, **extra_args)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/comfy/samplers.py", line 323, in forward
out = self.inner_model(x, sigma, cond=cond, uncond=uncond, cond_scale=cond_scale, cond_concat=cond_concat, model_options=model_options, seed=seed)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/comfy/k_diffusion/external.py", line 125, in forward
eps = self.get_eps(input * c_in, self.sigma_to_t(sigma), **kwargs)
File "/workspace/ComfyUI/comfy/k_diffusion/external.py", line 151, in get_eps
return self.inner_model.apply_model(*args, **kwargs)
File "/workspace/ComfyUI/comfy/samplers.py", line 311, in apply_model
out = sampling_function(self.inner_model.apply_model, x, timestep, uncond, cond, cond_scale, cond_concat, model_options=model_options, seed=seed)
File "/workspace/ComfyUI/comfy/samplers.py", line 289, in sampling_function
cond, uncond = calc_cond_uncond_batch(model_function, cond, uncond, x, timestep, max_total_area, cond_concat, model_options)
File "/workspace/ComfyUI/comfy/samplers.py", line 265, in calc_cond_uncond_batch
output = model_function(input_x, timestep_, **c).chunk(batch_chunks)
File "/workspace/ComfyUI/comfy/model_base.py", line 63, in apply_model
return self.diffusion_model(xc, t, context=context, y=c_adm, control=control, transformer_options=transformer_options).float()
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/comfy/ldm/modules/diffusionmodules/openaimodel.py", line 659, in forward
h = forward_timestep_embed(module, h, emb, context, transformer_options, output_shape)
File "/workspace/ComfyUI/comfy/ldm/modules/diffusionmodules/openaimodel.py", line 56, in forward_timestep_embed
x = layer(x, context, transformer_options)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/comfy/ldm/modules/attention.py", line 695, in forward
x = block(x, context=context[i], transformer_options=transformer_options)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/comfy/ldm/modules/attention.py", line 525, in forward
return checkpoint(self._forward, (x, context, transformer_options), self.parameters(), self.checkpoint)
File "/workspace/ComfyUI/comfy/ldm/modules/diffusionmodules/util.py", line 123, in checkpoint
return func(*inputs)
File "/workspace/ComfyUI/comfy/ldm/modules/attention.py", line 635, in _forward
x = self.ff(self.norm3(x)) + x
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/comfy/ldm/modules/attention.py", line 82, in forward
return self.net(x)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/container.py", line 217, in forward
input = module(input)
File "/workspace/ComfyUI/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "/workspace/ComfyUI/comfy/ldm/modules/attention.py", line 62, in forward
return x * F.gelu(gate) 
```

**Comment by cubiq** [comment]
each image in IPAdapter is 16 tokens. You are actively sending 256 tokens to the pipeline :) Also of course each image takes some ram. Honestly I'm surprised it works even with 8 images. You can try with the not-plus version that uses less tokens.

One thing you can try is to pre-encode the images with the new ipadapter encode node, but I'm not sure it would help much.

**Comment by sdexpanim** [comment]
Ah I see thank you, 

By the 256 tokens, you are referring to the 16 empty latent batch? I think I'm only sending one clip encoded image here. 

Thanks for the suggestion but yeah same issues when pre-encoding the image.

**Comment by cubiq** [comment]
oh yeah I'm sorry, I was reading your message on my email and didn't see the attached picture.

so yeah that has to be a memory limit. can you try to pre-encode the images?

**Comment by sdexpanim** [comment]
No worries at all, 
Yeah it looks like the same issue when pre-encoding the input image,
I guess trying to process a big batch of latents is causing IPadapter to use more vram? 


here's my workflow with the pre-encode:
![Screenshot 2023-10-01 000923](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/144590036/ae01fab3-8153-408d-b54d-3311e62e05dd)

**Comment by cubiq** [comment]
the workflow is correct, try to use the IPAdapter save embeds, and then the load embeds. At that point you can remove the clip encoder and all the images from your workflow. I would also stop Comfy so you are sure all the available ram is freed

![save](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/427614/652f14cc-38f4-4b9c-9050-1cfa51c640c6)

**Comment by sdexpanim** [comment]
Sorry yeah that makes sense, need to load the embeds in a separate workflow. 
Unfortunately same results.
![Screenshot 2023-10-01 002523](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/144590036/67431a25-a522-436d-bc59-0db25802a62c)

**Comment by cubiq** [comment]
I'll see if there's any further optimization I can do but I'm afraid we are reaching the limit of what we can do. thanks for your report anyway

**Comment by sdexpanim** [comment]
All good, yeah seems like the IPadapter model is a bit vram hungry for the time being, maybe in the future some optimisation will be made. 

thanks for your time and effort! Happy to mark this as resolved

**Comment by cubiq** [comment]
do you mind making a test with `--force-fp16` command line option?

**Comment by sdexpanim** [comment]
unfortunately the same issue. This is running with 16 empty latents at 768x768. Same on 24 empty latents at 512x768. 
I think it's using slightly less vram but still hits a cap for those sizes.

![Screenshot 2023-10-05 202819](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/144590036/211289d9-d030-452f-8ac4-0ed010992670)

**Comment by angeloshredder** [comment]
I don't think it's a IP Adapter problem, 
in my last workflow, i used 8 images as input for IP Adapter, 3 ControlNet , 3 KSamplers  at 60 steps dpmpp_3m_sde karras , outputs were 920x920, all this workflow needed 10.1/24GB Vram at max 
![00000-2023-10-05_174328](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/52110838/ef8a1211-761b-4fc7-87dc-7690d9cda156)

**Comment by cubiq** [comment]
comfyui recently changed memory management and I'm seeing a lot more OOM. The number of images in IPAdapter should have an impact on the computational time (more tokens, more time) and on the used RAM but not so much on VRAM. Using pre encoded images should spare about 1.5gb of VRAM.

PS: dpmpp_3m_sde is supposed to be used with exponential or sgm schedulers

**Comment by angeloshredder** [comment]
"PS: dpmpp_3m_sde is supposed to be used with exponential or sgm schedulers"
thanks for this information. 
I tried with an SD1.5 model, not sure about SDXL VRAM usage !!

**Comment by sdexpanim** [comment]
thanks @angeloshredder, I think your workflow is a bit different. My issue comes from using a large batch of empty latents. I'm only using 1 clip vision embedded image in the IPadatper model. Without IPadapter I can have huge batches of empty latents, like up to 128 before OOM, and I can use much higher resolutions, it just takes much longer to process. 

Large batch of empty latents is only really useful in animateDiff, which is my main usecase for IPadapter.

That's interesting @cubiq , I wonder if comfy is aware of any memory issues after said changes.

**Comment by sdexpanim** [comment]
I'm really confused honestly. I've spoken to others running IPadapter from this repo https://github.com/laksjdjf/IPAdapter-ComfyUI and they say they're using 10GB of VRAM for 768x768 at 16 latents for ipadapter. I tried that repo and my runpod 4090 goes over 24GB :( 

It doesn't make sense at all to me

**Comment by cubiq** [comment]
the models are the same, I doubt there's any difference in VRAM usage, my extension uses less RAM though.

it's hard to escalate without looking what they are actually doing. Maybe a workflow would help

**Comment by sdexpanim** [comment]
Ok, thanks again. I'll try get more info off them in the morning and do some more tests myself.

**Comment by cubiq** [comment]
I forgot to mention, it's worth trying to run comfy with `--disable-smart-memory` and if that's not enough also `--lowvram`

That should bring things back to before the update

**Comment by sdexpanim** [comment]
Thanks for all you effort on this, I don't know what happened but The VRAM issues have gone away. I switched to a different image for comfy on runpod and now never get OOM for ipadapter. Even at 768x768 at 32 latent frames. 

I tried it on your repo and the other one. With animate diff loader I can run many frames without running out. Without animateDiff loader however VRAM cap is hit. I heard from the developer this is due to how animate diff loader loads batches of latents into vram in "context sizes". 

So I think this issue is resolved for now. I'm not sure why I was running into issues with animateDiff + ipadapter before but I no longer have those issues. 

I don't think it makes a difference but for anyone else using the same setup. I was using the official comfy runpod image before and now switched to `Stable Diffusion Kohya_ss ComfyUI Ultimate` community one. 

Here's my workflow that works with ipadatper + animateDiff for anyone:

![ComfyUI_00303_](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/144590036/7c77f2ce-a7ed-442a-89b8-404f86898daf)
